cornac/models/cdl/cdl.py

```diff
@@ -1,144 +1,182 @@
+                                     initializer=self.init_params.get('U', None))
+
+            self.V = tf.get_variable(name='V', shape=[self.n_items, self.k], dtype=tf.float32,
+                                     initializer=self.init_params.get('V', None))
+
+
+        self.cdl_batch = tf.placeholder(dtype=tf.int32)
+        real_batch_size = tf.cast(tf.shape(self.text_input)[0], tf.int32)
+        V_batch = tf.reshape(tf.gather(self.V, self.cdl_batch), shape=[real_batch_size, self.k])
+
+        # The Weight and Bias for each layer in SDAE
+        L = len(self.layers)
+        weights, biases = dict(), dict()
+        with tf.variable_scope("SDAE_Variable"):
+            for i in range(L - 1):
+                weights[i] = tf.get_variable(name='W_{}'.format(i), dtype=tf.float32,
+                                             shape=(self.layers[i], self.layers[i + 1]))
+                biases[i] = tf.get_variable(name='b_{}'.format(i), dtype=tf.float32,
+                                            shape=(self.layers[i + 1]),
+                                            initializer=tf.zeros_initializer())
+
+        corrupted_text = tf.multiply(self.text_input, self.mask_input)
+
+        sdae_value, encoded_value = Model.sdae(corrupted_text, self.layers, weights, biases, self.dropout_rate)
+
+        # Generate loss function
+        loss_w_b = tf.constant(0, dtype=tf.float32)
+        for i in range(L - 1):
+            loss_w_b = tf.add(loss_w_b, tf.nn.l2_loss(weights[i]))
+
+        loss_1 = self.lambda_u * tf.nn.l2_loss(self.U) + self.lambda_w * loss_w_b
+        loss_2 = self.lambda_v * tf.nn.l2_loss(V_batch - encoded_value)
+        loss_3 = self.lambda_n * tf.nn.l2_loss(sdae_value - self.text_input)
+        loss_4 = tf.reduce_sum(tf.multiply(self.C_input,
+                                           tf.square(self.rating_input - tf.matmul(self.U, V_batch, transpose_b=True))))
+
+        self.loss = loss_1 + loss_2 + loss_3 + loss_4
+
+        # Generate optimizer
+        optimizer1 = tf.train.AdamOptimizer(self.lr)
+        optimizer2 = tf.train.AdamOptimizer(self.lr)
+
+        train_var_list1, train_var_list2 = [], []
+
+        for var in tf.trainable_variables():
+            if "CDL_Variable" in var.name:
+                train_var_list1.append(var)
+            elif "SDAE_Variable" in var.name:
+                train_var_list2.append(var)
+
+        gvs = optimizer1.compute_gradients(self.loss, var_list=train_var_list1)
+        capped_gvs = [(tf.clip_by_value(grad, -5., 5.), var) for grad, var in gvs]
+        self.opt1 = optimizer1.apply_gradients(capped_gvs)
+
+        gvs = optimizer2.compute_gradients(self.loss, var_list=train_var_list2)
+        capped_gvs = [(tf.clip_by_value(grad, -5., 5.), var) for grad, var in gvs]
+        self.opt2 = optimizer2.apply_gradients(capped_gvs)
+
+    # Stacked Denoising Autoencoder
+    @staticmethod
+    def sdae(X_corrupted, layer_sizes, weights, biases, dropout_rate):
```

Comment:
This can be a private method of Model class, which avoid passing layer_sizes and dropout_rate as arguments